File: src/main/java/itmo/is/project/repository/module/dock/DockingSpotRepository.java
package itmo.is.project.repository.module.dock;

import itmo.is.project.model.module.dock.DockingSpot;
import itmo.is.project.model.user.Spaceship;
import itmo.is.project.model.user.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface DockingSpotRepository extends JpaRepository<DockingSpot, Integer> {
    Page<DockingSpot> findAllByIsOccupiedTrue(Pageable pageable);

    Optional<DockingSpot> findFirstBySizeAndIsOccupiedFalse(Spaceship.Size size);

    Optional<DockingSpot> findBySpaceshipPilot(User pilot);

    boolean existsBySpaceship(Spaceship spaceship);
}

File: src/main/kotlin/itmo/isproject/repository/module/dock/DockingSpotRepository.kt
package itmo.isproject.repository.module.dock

import itmo.isproject.model.module.dock.DockingSpot
import itmo.isproject.model.user.Spaceship
import itmo.isproject.model.user.User
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository

@Repository
interface DockingSpotRepository : JpaRepository<DockingSpot, Int> {

    fun findAllByIsOccupiedTrue(pageable: Pageable): Page<DockingSpot>

    fun findFirstBySizeAndIsOccupiedFalse(size: Spaceship.Size): DockingSpot?

    fun findBySpaceshipPilot(pilot: User): DockingSpot?

    fun existsBySpaceship(spaceship: Spaceship): Boolean
}
